databroker/projector.py

```diff
@@ -137,54 +259,100 @@ def project_xarray(run: BlueskyRun, *args, projection=None, projection_name=None
+        in the passed-in projection. These are projections with "location": "linked"...note that
+        every xarray for a field froma given stream will contain a reference to the same set of configuration attrs
+        for as all fields from the same stream
+
+        Dataset
+            |_attrs
+                |_'projection_start_field': value
+            |_data
+                |_ 'projection_event_field': xarray
+                                                |_ attrs
+                                                    |_'projection_configuration_field': value
 
     Raises
     ------
     ProjectionError
     """
+    attrs = {}  # will populate the return Dataset attrs field
+    data_vars = {}  # will populate the return Dataset DataArrays
+    stream_configurations = {}  # will populate a collection of dicts of stream configurations
+
+    def metadata_cb(field, value):
+        attrs[field] = value
+
+    def event_configuration_cb(
+            projection_field,
+            stream,
+            config_index,
+            config_device,
+            config_field,
+            value):
+
+        if stream not in stream_configurations:
+            stream_configurations[stream] = []
+        if len(stream_configurations[stream]) == 0:
+            stream_configurations[stream].append({})
+        if config_device not in stream_configurations[stream][config_index]:
+            stream_configurations[stream][config_index][config_device] = {}
+
+        stream_configurations[stream][config_index][config_device][config_field] = value
+
+    def event_field_cb(projection_field,
+                       stream,
+                       field,
+                       xarray: xarray.DataArray):
+
+        if projection_field not in stream_configurations:
+            stream_configurations[stream] = []
+        # associate the stream configuration to the xarrays's atrtrs
+        xarray.attrs['configuration'] = stream_configurations[stream]
+        data_vars[projection_field] = xarray
     try:
-        if projection is None:
-            projection = get_run_projection(run, projection_name)
-        if projection is None:
-            raise ProjectionError("Projection could not be found")
-
-        attrs = {}  # will populate the return Dataset attrs field
-        data_vars = {}  # will populate the return Dataset DataArrays
-        for field_key, mapping in projection['projection'].items():
-            # go through each projection
-            projection_type = mapping['type']
-            projection_location = mapping.get('location')
-            projection_data = None
-            projection_linked_field = mapping.get('field')
-
-            # single value data that will go in the top
-            # dataset's attributes
-            if projection_location == 'configuration':
-                attrs[field_key] = run.metadata['start'][projection_linked_field]
-                continue
+        projector = Projector(
```

Comment:
Is this `try` block intended to be part of the `event_field_cb` method? As written it runs once at class definition time. If this "works" it might indicate a need for another unit test.